[WebGPU] Support writing implementations of C++ members in Swift extensions

emw-apple opened the pull request:
#### d8a9b5a6e9815df71cc3a821ca8068bf425cd163
<pre>
[WebGPU] Support writing implementations of C++ members in Swift extensions
<a href="https://bugs.webkit.org/show_bug.cgi?id=283151">https://bugs.webkit.org/show_bug.cgi?id=283151</a>
<a href="https://rdar.apple.com/problem/139928444">rdar://problem/139928444</a>

Reviewed by Geoffrey Garen.

Swift does not currently have an equivalent of ObjC interop&apos;s
@implementation for C++ classes, i.e. a mechanism to implement a member
function of a C++ class in Swift. With some careful use of preprocessor
macros, provide an interface for shimming our way to a similar end. The
goal is for WebGPU member functions to be able to be written as Swift
extensions, with minimal overhead that can be cleaned up once
first-class support is available.

This is highly desirable for WebGPU, since it gives us a pathway to
reimplement bits of the project without affecting the overall WebGPU
class hierarchy, or resorting to awkward C-style object oriented
programming, where we only write free functions and have no
encapsulation rules to guard member access.

First, add a PUBLIC_IN_WEBGPU_SWIFT macro to modify encapsulation rules.
Swift&apos;s C++ interop layer only grants access to public members, even
in extensions. PUBLIC_IN_WEBGPU_SWIFT expands to `public` when compiling
WebGPU&apos;s swift module only, so it can be used to break encapsulation for
implementation purposes while preventing clients from accesing fields.

Second, introduce a pair of SWIFTCXX_THUNK macros to mark C++ member
functions that are implemented in Swift code. HAS_SWIFTCXX_THUNK
annotates a header declaration and indicates that the function is
refined by Swift. DEFINE_SWIFTCXX_THUNK generates a C++
member function that calls a Swift thunking function, which itself calls
the real extension member. The Swift thunk must be manually written. I
intend to add a Swift macro to generate the thunk in the future.

Putting it all together, reimplement WebGPU::Buffer::copy, attempting to
use idiomatic and memory-safe Swift. As a drive-by fix, rename it to
&quot;copyFrom&quot; to better reflect semantics. It works like this:

1. Buffer.h declares `copyFrom` as:

       void copyFrom(const std::span&lt;const uint8_t&gt;, const size_t offset) HAS_SWIFTCXX_THUNK;

   The NS_REFINED_FOR_SWIFT prevents Swift code from binding to this symbol
   and needlessly jumping through the thunks.

2. Buffer.mm implements the C++ thunk by providing the member name, its
   return type, and its argument types:

       SWIFTCXX_THUNK(WebGPU::Buffer, copyFrom, void, const std::span&lt;const uint8_t&gt;, const size_t);

3. The C++ thunk calls into a hand-written Swift thunk (this is what I
   intend to macro-ify in the future):

       public func Buffer_copyFrom_thunk(_ buffer: WebGPU.Buffer, from data: SpanConstUInt8, offset: Int) {
           buffer.copy(from: data, offset: offset)
       }

4. The Swift thunk calls into the actual extension code:

       extension WebGPU.Buffer {
           func copy(from data: SpanConstUInt8, offset: Int) {
               ...
           }
       }

   ...which has access to all members of WebGPU::Buffer that are marked
   PUBLIC_IN_WEBGPU_SWIFT.

These changes are all in service of the &quot;copy&quot; to &quot;copyFrom&quot; rename:

* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
(WebCore::GPUBuffer::internalUnmap):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
(WebCore::WebGPU::BufferImpl::copyFrom):
(WebCore::WebGPU::BufferImpl::copy): Deleted.
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBuffer.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::copy):

The rest:

* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferCopy):
(WebGPU::Buffer::copy): Deleted.
* Source/WebGPU/WebGPU/Buffer.swift:
(WebGPU.bufferContents):
(WebGPU.copy(from:offset:)):
(Buffer_copyFrom_thunk(_:from:offset:)):
(copySpan(_:source:)): Deleted.
* Source/WebGPU/WebGPU/CommandsMixin.h:
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
* Source/WebGPU/WebGPU/SwiftCXXThunk.h: Added.
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebGPU/WebGPU/WebGPUSwift.h: Removed. I thought I would need it
  in &lt;<a href="https://commits.webkit.org/286661@main">https://commits.webkit.org/286661@main</a>&gt;, but WebGPUExt.h is a
  better home for declarations that need to be included everywhere.

Canonical link: <a href="https://commits.webkit.org/286822@main">https://commits.webkit.org/286822@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/903bef1b8d21e44a2fc78488eff40f06032e7eb9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77194 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56229 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81757 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/28476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79311 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65377 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4525 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60480 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/28476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80261 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50441 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66255 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40782 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47844 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23753 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26799 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68958 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24088 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83181 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4574 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68758 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4730 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66228 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68014 "Found 1 new API test failure: /TestWebKit:WebKit.FailedLoad (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16975 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/12001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/10091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4521 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/4540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/7975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6299 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebGPU/WebGPU/WebGPUSwift.h:
```diff
@@ -25,11 +25,16 @@
+#if defined(__swift__) && __swift__ && \
+    defined(__WEBGPU__) && __WEBGPU__ && \
+    defined(ENABLE_WEBGPU_SWIFT) && ENABLE_WEBGPU_SWIFT
+#define WebGPUSwift_protected   public
```
I'm struggling with this naming because it tells the reader where it will be used, and not what it's going to do.

We could make it more verbose:
`PROTECTED_BUT_PUBLIC_IN_WEBGPU_SWIFT:`

Or use it like this:
```
protected:
PUBLIC_IN_WEBGPU_SWIFT:

```
Or like this:

`protected: PUBLIC_IN_WEBGPU_SWIFT:`

I think I like the last version best? It's a bit weird, but it says what it does.

emw-apple replied:
That makes sense, you've convinced me too. Though I think it needs to be

`protected: PUBLIC_IN_WEBGPU_SWIFT`

(no trailing colon) so that the macro can expand to nothing outside of the Swift target's compilation.

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.swift:
```diff
@@ -23,7 +23,24 @@
+    }
+}
+
+// FIXME(emw): Find a way to generate thunks like these, maybe via a macro?
```
We could generate these, but only if we dropped the name prefix on all function arguments.

emw-apple replied:
I'll post my macro implementation next week — it generates working code, though it doesn't get emitted to the C++ header due rdar://139984562.

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -28,13 +28,17 @@
+#if ENABLE(WEBGPU_SWIFT)
+#import "WebGPUSwiftInternal.h"
+
+SWIFTCXX_THUNK(WebGPU::Buffer, copy, void, const std::span<const uint8_t>, const size_t);
```
I would call this DEFINE_SWIFTCXX_THUNK. In case there's ever a DECLARE partner name.

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.h:
```diff
@@ -109,7 +109,7 @@ class Buffer : public WGPUBufferImpl, public RefCounted<Buffer>, public CanMakeW
     void indirectBufferInvalidated();
 #if ENABLE(WEBGPU_SWIFT)
-    void copy(const std::span<const uint8_t>, const size_t offset);
+    void copy(const std::span<const uint8_t>, const size_t offset) NS_REFINED_FOR_SWIFT;
```
Should we make a DECLARE_SWIFT_CXX_THUNK to do this declaration?

geoffreygaren replied:
DECLARE_SWIFT_CXX_THUNK can do nothing if !ENABLE(WEBGPU_SWIFT) and also add NS_REFINED_FOR_SWIFT at the end.

geoffreygaren replied:
While we're here, is this a copyTo or a copyFrom? Let's update the name.

emw-apple replied:
Since the C++ declaration is the interface, I think it makes sense to write it out. But I added a `HAS_SWIFTCXX_THUNK` which expands to NS_REFINED_FOR_SWIFT, to make our intent clearer. That's kind of like the `NS_SWIFT_NAME` attribute that Swift-ObjC interop uses.

nmahendru commented on Source/WebGPU/WebGPU/Buffer.h:
```diff
@@ -123,7 +123,9 @@ class Buffer : public WGPUBufferImpl, public RefCounted<Buffer>, public CanMakeW
     void incrementBufferMapCount();
     void decrementBufferMapCount();
 
+WebGPUSwift_private:
```
How about we used private_begin(public:)/private_end(private:) instead of WebGPUSwift_private:/private:.
When Swift team comes around providing us with a better mechanism then we define all these macros to "" and we are done.. else we will be left with redundant `private:` specifiers which we manually have to go remove, although they are harmless:
```
private:

...
...

private:

...
...
private:

.....
```

geoffreygaren replied:
If we go with `protected: PUBLIC_IN_WEBGPU_SWIFT`, then we can just grep and delete all instances of PUBLIC_IN_WEBGPU_SWIFT once we have compiler support.

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.h:
```diff
@@ -123,7 +123,9 @@ class Buffer : public WGPUBufferImpl, public RefCounted<Buffer>, public CanMakeW
     void incrementBufferMapCount();
     void decrementBufferMapCount();
 
+PUBLIC_IN_WEBGPU_SWIFT
```
This is fine, but I think we can also have an idiom going forward where we always write PUBLIC_IN_WEBGPU_SWIFT right next to the existing qualifier. 

The net effect will make more things public in WebGPU Swift, but it will be cleaner, and once we commit to exposing private stuff, I don’t see any value in being selective about it.

emw-apple replied:
This gave me the idea that we can define PUBLIC_IN_WEBGPU_SWIFT to be `: public` with a leading colon, so that it must be written after another access specifier. And we can implement your original syntax, where it's more obviously acting as an access specifier:

```
private PUBLIC_IN_WEBGPU_SWIFT:
...
```

geoffreygaren reviewed (approved):
r=me

emw-apple committed:
[WebGPU] Support writing implementations of C++ members in Swift extensions
https://bugs.webkit.org/show_bug.cgi?id=283151
rdar://problem/139928444

Reviewed by Geoffrey Garen.

Swift does not currently have an equivalent of ObjC interop's
@implementation for C++ classes, i.e. a mechanism to implement a member
function of a C++ class in Swift. With some careful use of preprocessor
macros, provide an interface for shimming our way to a similar end. The
goal is for WebGPU member functions to be able to be written as Swift
extensions, with minimal overhead that can be cleaned up once
first-class support is available.

This is highly desirable for WebGPU, since it gives us a pathway to
reimplement bits of the project without affecting the overall WebGPU
class hierarchy, or resorting to awkward C-style object oriented
programming, where we only write free functions and have no
encapsulation rules to guard member access.

First, add a PUBLIC_IN_WEBGPU_SWIFT macro to modify encapsulation rules.
Swift's C++ interop layer only grants access to public members, even
in extensions. PUBLIC_IN_WEBGPU_SWIFT expands to `public` when compiling
WebGPU's swift module only, so it can be used to break encapsulation for
implementation purposes while preventing clients from accesing fields.

Second, introduce a pair of SWIFTCXX_THUNK macros to mark C++ member
functions that are implemented in Swift code. HAS_SWIFTCXX_THUNK
annotates a header declaration and indicates that the function is
refined by Swift. DEFINE_SWIFTCXX_THUNK generates a C++
member function that calls a Swift thunking function, which itself calls
the real extension member. The Swift thunk must be manually written. I
intend to add a Swift macro to generate the thunk in the future.

Putting it all together, reimplement WebGPU::Buffer::copy, attempting to
use idiomatic and memory-safe Swift. As a drive-by fix, rename it to
"copyFrom" to better reflect semantics. It works like this:

1. Buffer.h declares `copyFrom` as:

       void copyFrom(const std::span<const uint8_t>, const size_t offset) HAS_SWIFTCXX_THUNK;

   The NS_REFINED_FOR_SWIFT prevents Swift code from binding to this symbol
   and needlessly jumping through the thunks.

2. Buffer.mm implements the C++ thunk by providing the member name, its
   return type, and its argument types:

       SWIFTCXX_THUNK(WebGPU::Buffer, copyFrom, void, const std::span<const uint8_t>, const size_t);

3. The C++ thunk calls into a hand-written Swift thunk (this is what I
   intend to macro-ify in the future):

       public func Buffer_copyFrom_thunk(_ buffer: WebGPU.Buffer, from data: SpanConstUInt8, offset: Int) {
           buffer.copy(from: data, offset: offset)
       }

4. The Swift thunk calls into the actual extension code:

       extension WebGPU.Buffer {
           func copy(from data: SpanConstUInt8, offset: Int) {
               ...
           }
       }

   ...which has access to all members of WebGPU::Buffer that are marked
   PUBLIC_IN_WEBGPU_SWIFT.

These changes are all in service of the "copy" to "copyFrom" rename:

* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
(WebCore::GPUBuffer::internalUnmap):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
(WebCore::WebGPU::BufferImpl::copyFrom):
(WebCore::WebGPU::BufferImpl::copy): Deleted.
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBuffer.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::copy):

The rest:

* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferCopy):
(WebGPU::Buffer::copy): Deleted.
* Source/WebGPU/WebGPU/Buffer.swift:
(WebGPU.bufferContents):
(WebGPU.copy(from:offset:)):
(Buffer_copyFrom_thunk(_:from:offset:)):
(copySpan(_:source:)): Deleted.
* Source/WebGPU/WebGPU/CommandsMixin.h:
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
* Source/WebGPU/WebGPU/SwiftCXXThunk.h: Added.
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebGPU/WebGPU/WebGPUSwift.h: Removed. I thought I would need it
  in <https://commits.webkit.org/286661@main>, but WebGPUExt.h is a
  better home for declarations that need to be included everywhere.

Canonical link: https://commits.webkit.org/286822@main
